driver/transform-dsl.md

```diff
@@ -0,0 +1,549 @@
+    // store the value
+    Var("x"),
+)
+```
+
+### Objects
+
+#### Empty object
+
+To represent an empty object, `Obj{}` can be used.
+
+#### Object or nil
+
+To check any node for nil or specific rule `Opt` is used:
+
+```
+Opt("is_nil", Obj{ ... })
+```
+
+`Opt` needs a named variable to store it's state ("is_nil" in this case).
+
+#### All fields must exist
+
+For object that always have a specific set of fields, `Obj` operation
+can be used:
+
+```golang
+Obj{
+    "type": String("Ident"),
+    "name": Var("name"),
+}
+```
+
+Since this operation is represented by a map, order of operations is not
+preserved. Operations will be executed in order of field name sorting.
+This might affect operations that requres a variable to be defined, such
+as `Lookup` and `If`. To control execution order, use `Fields`.
+
+#### Specific field options
+
+To get more control of how object fields are checked, use `Fields`:
+
+```golang
+Fields{
+    {Name: "type", Op: String("If")},
+    {Name: "then": Op: Var("then")},
+    {Name: "else": Op: Var("then"), Optional: "else_exists"},
+}
+```
+
+For each optional field a variable should be provided to store the state.
+Optional field means that they may not exist. For checking nil fields see
+`Opt` operation.
+
+#### Skip unknown fields
+
+To skip unknown fields in object `Part` is used:
+
+```golang
+Part("other", Obj{
+    "name": Var("x"),
+})
+```
+
+`Part` will store all unhandled fields in `other` variable. It usually
+requires to have `Part` on both sides of a mapping.
+
+### Arrays
+
+#### Fixed number of elements
+
+To represet an array with a specific number of elements, `Arr` can be used:
+
+```golang
+Arr(
+    // element 0, fixed
+    Int(42),
+
+    // element 1, saved to a variable
+    Var("x"),
+)
+```
+
+A helper is available for arrays with exactly one element: `One(Int(42))`.
+
+#### Any number of elements
+
+To apply an operation to each array element, use `Each`:
+
+```golang
+Each("elems",
+    Obj{
+        "name": Var("x"),
+    },
+)
+```
+
+Note that `Each` requires a named variable to store elements. This also
+means that `Each` should always be on both side of a mapping.
+
+`Each` allows an array to be empty or nil.
+
+Operation that is passed to `Each` will receive a clean state, thus
+it will not be able to access any variables from the parent. It also
+means that variable names inside this operation will not collide with
+the variables defined by parent transformation.
+
+#### Any number of elements followed by a fixed "suffix"
+
+To match arrays with a specific set of nodes at the end, use `Append`:
+
+```golang
+Append(
+    Var("x"),
+    Arr(Int(42)),
+)
+```
+
+`Append` will verify that node is an array and will filter out "suffix"
+nodes before passing it to primary operation (`Var("x")` in this case).
+
+When used to costruct the tree, it will append "suffix" nodes to
+the end of an array defined by sub-operation.
+
+As in Go, `Append` will not affect variable "x".
```

Comment:
Not about this specific line: missing some doc and examples for `LookupArrOpVal` and friends.

Reply:
What do you think about this?